[android] - Refactor dependencies, introduce focused components

tobrun opened the pull request:
WIP to refactor all gesture related code into a separate wrapper.
- [x] extract code from MapView and make it work
- [x] minimize internal dependencies on GestureDectector
- [x] remove MapView related methods, talk to NativeMapView directly.

**update**: intent has changed from initial creation see below

ivovandongen commented:
@tobrun There are indeed a lot of inter-dependecies. And also some 2-way dependencies. It might be helpful for the clarity of the architecture to abstract some things a little bit more. For example, 

- The `MapGestureDetector` has a dependency on MapView, which actually is only passed to the `ZoomButtonsController` as far as I see which only expects a `View`, not a concrete class like `MapView`.  
- In case of the `MapboxMap` dependency, it's actually the `Transform`, `Projection` and `UISettings` you're after.
- The `NativeMapView` dependency is needed for 2 things; to get the pixelRatio/pitch and to control and to effect the gestures on the Map. You might want to abstract the second repsonsibility in an Interface.

There might be a couple of places where interfaces/choosing the highest applicable abstraction might help to clear up the dependencies a bit.

tobrun commented:
> The MapGestureDetector has a dependency on MapView, which actually is only passed to the ZoomButtonsController as far as I see which only expects a View, not a concrete class like MapView.

Thinking about this again, the zoombutton controller is more like a widget shown on top of the map, I moved this back to MapView with the other widgets. These will be extracted as part of #6570.

>    In case of the MapboxMap dependency, it's actually the Transform, Projection and UISettings you're after.

I tried to limit the amount of arguments but not assigning mapbox to field in the MapGestureDetector. I have now changed this to take each dependency seperatly. 

 
>  The NativeMapView dependency is needed for 2 things; to get the pixelRatio/pitch and to control and to effect the gestures on the Map. You might want to abstract the second repsonsibility in an Interface.

I'm now working to remove the NativeMapView in total, the second responsibility should be made a part of the transform class since they transform the map.

Thanks for some initial :eyes:, will ping again for second round.

tobrun commented:
The goal of this PR has diverged from it's initial intent, changes made cascaded in changes elsewhere. This PR is now more focused on a general refactor and taking apart the big dependencies into smaller single focused dependencies. This PR adds a lot of todo to the code but current state is acceptable to merge. (we are only missing the snapshot feature, the jni code needs to be refactored and can be picked up with #6238). Focus of Review would be MapboxMap, MapGestureDetector, MapKeyListener, Transform, TrackingSettings and UiSettings. To make the refactor possible I had to remove some unit tests.

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerViewManager.java:
```diff
@@ -31,27 +32,33 @@
-        this.mapboxMap = mapboxMap;
+    public MarkerViewManager(@NonNull ViewGroup container) {
+        this.markerViewContainer = container;
         this.markerViewAdapters = new ArrayList<>();
```
The initializations that have no dependencies might as well be done in the field declaration. That also enables us to make them final and maybe skip some null checks here and there.

tobrun replied:
:white_check_mark:

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java:
```diff
@@ -33,29 +36,39 @@
-        this.iconManager = iconManager;
-        this.infoWindowManager = manager;
+        this.iconManager = new IconManager(nativeMapView);
+        this.infoWindowManager = new InfoWindowManager();
```
Initialize in field declaration

tobrun replied:
:white_check_mark:

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java:
```diff
@@ -33,29 +36,39 @@
         }
     }
 
+    // TODO refactor MapboxMap out for Projection and Transform
```
Let's note these things on #6912 as well. I'm planning to pick that up soon.

tobrun replied:
https://github.com/mapbox/mapbox-gl-native/issues/6912#issuecomment-264445033

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapGestureDetector.java:
```diff
@@ -0,0 +1,598 @@
+     */
+    boolean onGenericMotionEvent(MotionEvent event) {
+        // Mouse events
+        //if (event.isFromSource(InputDevice.SOURCE_CLASS_POINTER)) { // this is not available before API 18
```
Stray comment?

tobrun replied:
it seems not.. after that line it indicates:  `// this is not available before API 18`, so it's  there waiting for us to bump the minSDK requirments? IDK, this code is from before my time.

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java:
```diff
@@ -180,13 +137,14 @@ public MapView(@NonNull Context context, @Nullable MapboxMapOptions options) {
 
+        // TODO distill into singular purpose methods/classes
         initialLoad = true;
         onMapReadyCallbackList = new ArrayList<>();
```
Initialize in declaration

tobrun replied:
:white_check_mark:

tobrun commented:
I have worked in the requested changes and working now on getting the MapboxMapTest back up and running since this contains our public API

tobrun commented:
I got the important tests back to run by using mockito 2 configuration for mocking final classes. Though I wasn't able to bring the MapboxMap test since they rely on NativeMapView (which relies on loading the native library). I'm thinking of porting those tests to an instrumentation test equivalent. I think we are covered for now. @ivovandongen can you do another review round?

ivovandongen commented:
> I wasn't able to bring the MapboxMap test since they rely on NativeMapView (which relies on loading the native library). I'm thinking of porting those tests to an instrumentation test equivalent. I think we are covered for now. 

Let's open a follow-up ticket for that then.

ivovandongen commented on platform/android/MapboxGLAndroidSDKTestApp/src/test/java/com/mapbox/mapboxsdk/maps/MapboxMapTest.java:
```diff
@@ -31,512 +34,482 @@
 import static org.mockito.Mockito.when;
 
 public class MapboxMapTest {
-
```
Why not just use @ignore on the class for now? Or does that make it crash. Makes it more visible that we need to do some work here.

tobrun commented:
@ivovandongen I have actually now have ported all the unit MapboxMapTest to an instrumentation equivalent. Can you review?

tobrun commented:
I'm running the Android job multiple times on Bitrise to ensure we aren't introducing any flaky tests.

tobrun commented:
Not seeing any immediate fixes, going to ignore the MapboxMapTest and create a follow up ticket.
